client/blocks/upload-image/index.jsx

```diff
@@ -0,0 +1,199 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import path from 'path';
+import Gridicon from 'gridicons';
+import { noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { ALLOWED_FILE_EXTENSIONS } from './constants';
+import { AspectRatios } from 'state/ui/editor/image-editor/constants';
+import Dialog from 'components/dialog';
+import FilePicker from 'components/file-picker';
+import {
+	resetAllImageEditorState
+} from 'state/ui/editor/image-editor/actions';
+import Spinner from 'components/spinner';
+import ImageEditor from 'blocks/image-editor';
+import DropZone from 'components/drop-zone';
+
+class UploadImage extends Component {
+	state = {
+		isEditingImage: false,
+		uploadedImage: null,
+		editedImage: null,
+	};
+
+	static propTypes = {
+		isUploading: PropTypes.bool,
+		// Additional props passed to ImageEditor component. See blocks/image-editor.
+		imageEditorProps: PropTypes.object,
+		texts: PropTypes.object,
+		onImageEditorDone: PropTypes.func,
+		additionalImageEditorClasses: PropTypes.string,
+		additionalClasses: PropTypes.string,
+	};
+
+	static defaultProps = {
+		imageEditorProps: {
+			allowedAspectRatios: [ AspectRatios.ASPECT_1X1 ],
+		},
+		texts: {},
+		backgroundContent: null,
+		onImageEditorDone: noop,
+		isUploading: false,
+	};
+
+	onReceiveFile = ( files ) => {
+		const extension = path.extname( files[ 0 ].name )
+			.toLowerCase()
+			.substring( 1 );
+
+		if ( ALLOWED_FILE_EXTENSIONS.indexOf( extension ) === -1 ) {
+			return;
+		}
+
+		const imageObjectUrl = URL.createObjectURL( files[ 0 ] );
+
+		this.setState( {
+			isEditingImage: true,
+			uploadedImage: imageObjectUrl
+		} );
+	};
+
+	onImageEditorDone = ( error, imageBlob ) => {
+		this.hideImageEditor();
+
+		if ( error ) {
+			return;
+		}
+
+		this.setState( { editedImage: URL.createObjectURL( imageBlob ) } );
```

Comment:
Since `this.hideImageEditor` is called just a few lines above this first, we don't need to do anything special here because we set `isEditingImage` to `false` there.